### PR TITLE
S3 max redirects

### PR DESCRIPTION
**Pull request by ianton-ru**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Add limit for http redirects in request to S3 storage ('s3_max_redirects')

Detailed description / Documentation draft:

Add limit for http redirects in request to S3 storage ('s3_max_redirects'). To disable redirects set 's3_max_redirects' to 0. By default up to 10 redirects allowed.
Issue: https://github.com/ClickHouse/ClickHouse/issues/12531

### CODE REVIEW

File: src/IO/S3/PocoHTTPClient.cpp [hunk 1]
```diff
@@ -83,6 +86,7 @@ PocoHTTPClient::PocoHTTPClient(const PocoHTTPClientConfiguration & clientConfigu
           Poco::Timespan(clientConfiguration.httpRequestTimeoutMs * 1000) /// receive timeout.
           ))
     , remote_host_filter(clientConfiguration.remote_host_filter)
+    , global_context(clientConfiguration.global_context)
```
Comment (alesapin) [inline src/IO/S3/PocoHTTPClient.cpp:89]: Maybe we can pass only settings without the whole context?

Comment (ianton-ru) [reply src/IO/S3/PocoHTTPClient.cpp]: Fixed

**Comment by alesapin** [comment]
`01035_lc_empty_part_bug` Probably fixed in https://github.com/ClickHouse/ClickHouse/pull/17311

**Comment by alesapin** [comment]
`test_storage_rabbitmq/test.py::test_rabbitmq_csv_with_delimiter` also fixed in #17311

**Comment by alesapin** [comment]
` Stress test (memory) — Server failed to start ` -- not related to changes

File: src/IO/S3/PocoHTTPClient.cpp [hunk 2]
```diff
@@ -50,9 +50,11 @@ namespace DB::S3
 
 PocoHTTPClientConfiguration::PocoHTTPClientConfiguration(
         const Aws::Client::ClientConfiguration & cfg,
-        const RemoteHostFilter & remote_host_filter_)
+        const RemoteHostFilter & remote_host_filter_,
+        unsigned int s3_max_redirects_)
```
Comment (alesapin) [inline src/IO/S3/PocoHTTPClient.cpp:54]: unsigned int -- quite strange to use this type, but ok, doesn't affect anything.

**Pull request merged**